src/backend/storage/file/buffile.c

```diff
@@ -998,6 +1000,34 @@ BufFilePledgeSequential(BufFile *buffile)
 
 #define BUFFILE_ZSTD_COMPRESSION_LEVEL 1
 
+typedef struct {
+	ResourceOwner oldOwner;
+	ResourceOwner newOwner;
+} ResourceOwnerContext;
```

Comment:
I meant such structure:
```c
typedef struct {
	ResourceOwner owner;
	MemoryContext context;
} ResourceOwnerMemoryContext;
```

Reply:
Could you give an example of how we should use `MemoryContext` in that case? Do you mean that we should use `MemoryContextAlloc` for memory allocation? If so, why do we need `MemoryContext` field in the struct if in code `TopMemoryContext` is used?

Reply:
> we should use `MemoryContextAlloc` for memory allocation?

Yes, it's possible.

Reply:
> why do we need `MemoryContext` field in the struct if in code `TopMemoryContext` is used?

You can do this:
```c
... = MemoryContextAlloc(TopMemoryContext, ...
```

Reply:
In that case, we don't need to pass any `MemoryContext`. Also, if we switch resource owner, shouldn't we switch it back? I mean that we need both old owner and new owner

Reply:
> if we switch resource owner, shouldn't we switch it back?

Yes!

Reply:
> we need both old owner and new owner

No!
1) save current resource owner to local variable
2) set current resource owner from context
3) doing something (palloc)
4) restore current resource owner from local variable

Reply:
Please, take a look, is that what you meant?

Reply:
> is that what you meant?

Yes, but with one member then the structure is not needed at all!

Reply:
Of course, I just wanted to make sure that it's what you meant